corehq/apps/app_manager/app_schemas/form_metadata.py

```diff
@@ -123,3 +130,87 @@ def get_app_summary_formdata(domain, app, include_shadow_forms=True):
     """Returns formdata formatted for the app summary
     """
     return _AppSummaryFormDataGenerator(domain, app, include_shadow_forms).generate()
+
+
+class AppDiffGenerator(object):
+    def __init__(self, app1, app2):
+        self.first = get_app_summary_formdata(app1.domain, app1)[0]
+        self.second = get_app_summary_formdata(app2.domain, app2)[0]
+
+        self._populate_id_caches()
+        self._mark_removed_items()
+        self._mark_added_items()
+
+    def _populate_id_caches(self):
+        self._first_ids = set()
+        self._first_questions_by_id = defaultdict(dict)
+        self._second_ids = set()
+        self._second_questions_by_id = defaultdict(dict)
+
+        for module in self.first:
+            self._first_ids.add(module['id'])
+            for form in module['forms']:
+                self._first_ids.add(form['id'])
+                for question in form['questions']:
+                    self._first_questions_by_id[form['id']][question['value']] = question
+
+        for module in self.second:
+            self._second_ids.add(module['id'])
+            for form in module['forms']:
+                self._second_ids.add(form['id'])
+                for question in form['questions']:
+                    self._second_questions_by_id[form['id']][question['value']] = question
+
+    def _mark_removed_items(self):
+        for module in self.first:
+            if module['id'] not in self._second_ids:
```

Comment:
Confused: `id` is the index, right? Shouldn't this look at unique id? It seems like the only way you'd mark a module as removed is if one app has a different number of modules than the other.

Reply:
`id` is actually the `unique_id`... 
https://github.com/dimagi/commcare-hq/blob/fr%2Fform-diffs/corehq/apps/app_manager/app_schemas/form_metadata.py#L125
Maybe that is worth changing here? It's been this way in the app summary for ever, but could be worth a refactor for consistency sake.

Reply:
If it's not a big increase in scope, it'd be a good improvement. Checking whether `id` is index or unique id is something I do fairly often in app manager.

I wish we could do `id` vs `index`, but changing the model properties would be unreasonable, so using `id` and `unique_id` consistently everywhere would be the next best thing.

Reply:
:+1: I will do this in a separate PR

Reply:
https://github.com/dimagi/commcare-hq/pull/24136